File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/kth_smallest_number_quick_select.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <vector>

using namespace std;

int rand_in(int low, int high) {
  random_device rd;
  mt19937 engine(rd());
  uniform_int_distribution dist(low, high);

  return dist(engine);
}

void swap(vector<int> &a, int x, int y) {
  int t = a[x];
  a[x] = a[y];
  a[y] = t;
}

int partition(vector<int> &a, int low, int high) {
  int pivot = a[low];
  int j = low;

  for (int i = low + 1; i <= high; i++) {
    if (a[i] <= pivot) {
      j++;

      if (i != j) {
        swap(a, i, j);
      }
    }
  }

  if (low != j) {
    swap(a, low, j);
  }

  return j;
}

int quick_select(vector<int> &a, int low, int high, int k) {
  int r = rand_in(low, high);
  swap(a, low, r);

  int mid = partition(a, low, high);
  if (mid == k) {
    return a[mid];
  } else if (mid < k) {
    return quick_select(a, mid + 1, high, k);
  }
  return quick_select(a, low, mid - 1, k);
}

// O(n2) / O(n)
int kth_smallest_number(vector<int> &a, int k) {
  return quick_select(a, 0, a.size() - 1, k - 1);
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << kth_smallest_number(a, k) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/KthSmallestNumberQuickSelect.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KthSmallestNumberQuickSelect {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.println(kthSmallest(a, k - 1));
      }
    }
  }

  // O(n2) / O(n)
  public static int kthSmallest(int[] a, int k) {
    return quickSelect(a, 0, a.length - 1, k);
  }

  private static int quickSelect(int[] a, int low, int high, int k) {
    int r = ThreadLocalRandom.current().nextInt(high - low + 1) + low;
    swap(a, low, r);

    int mid = partition(a, low, high);
    if (mid == k) {
      return a[mid];
    } else if (mid < k) {
      return quickSelect(a, mid + 1, high, k);
    }
    return quickSelect(a, low, mid - 1, k);
  }

  private static int partition(int[] a, int low, int high) {
    int pivot = a[low];
    int j = low;

    for (int i = low + 1; i <= high; i++) {
      if (a[i] <= pivot) {
        j++;

        if (i != j) {
          swap(a, i, j);
        }
      }
    }

    if (low != j) {
      swap(a, low, j);
    }

    return j;
  }

  private static void swap(int[] a, int x, int y) {
    int t = a[x];
    a[x] = a[y];
    a[y] = t;
  }
}